Add minimal interface to report Location errors in Astlib

NathanReb opened the pull request:
Given we'll need to be compatible with OCaml 4.02 and up to 4.09 and they have different interfaces even for the `raise_errorf` function I went for an abstract error type and a `raise` function so that it's easy to extend it without breaking compatibility and there is a single way to build and raise location errors.

We can then wrap this into convenience functions in `ppx` itself or in a sub library that can be shared by `ppx`, `ppx_metaquot` and `ppx_view`.

Adding optional arguments to the `make` function might not be considered strictly speaking non-breaking so we could go for a bunch of `set_x : x -> t -> t` functions.

I'm not 100% convinced by this approach so I'm happy to change it if you have any suggestions.

Let me know what you think!

NathanReb committed:
Add minimal interface to report Location errors in Astlib

ceastlund commented:
I think adding new optional arguments to "make" is fine, we can document in the interface that those might be added over time.

Why do we support calling make without a message string? Especially when there's no location argument, an empty error does not seem like a useful thing to support. Is there some existing use case for messageless errors that we're trying to preserve?

NathanReb commented:
My reasoning about the optional message is that in recent versions of ocaml, messages aren't passed around as strings but as formatters which allows for finer formatting of the message.

Making `msg` optional would allow to extend the make function with a `fmt_msg` argument to support this feature of newer compilers and seemed better than having a single `fmt_msg` which behaviour would depend on the OCaml version you're using and eventually result in different error messages.

It might be better to extract an abstract `msg` type with different constructors `raw_string_msg : string -> msg`, `fmt_msg: (Format.formatter -> unit) -> msg`, etc. and make it an mandatory argument of `make` makes for a better API here.

ceastlund commented:
Now we're adding two new types and multiple layers of nested applications to raise errors:

Error.raise (Error.make (Message.make ...))

I think it would be more straightforward to just start with one function to raise string-based errors, and in whatever compiler version we want to introduce formatters, add a second function to raise formatter-based errors. That way error code won't get more and more verbose over time.

I also think we might want to make loc required, and have people manually write Location.none where they want no location, rather than making it easy to forget to pass a location argument.

NathanReb commented:
I went for this kind of API because @diml mentioned that he wanted to avoid having too many ways to report errors in `Astlib` so that it doesn't end up like `Location`. I personally don't have strong opinions on the matter and I agree it makes for a too complex API to use if we throw too many types and smart constructors. My intent was to provide the minimal and easiest to extend API in `Astlib` and add proper wrappers in `ppx` where we can make opinionated choices about what's a good way to report errors in a ppx rewriter.

I'm fine with several raise-like functions as well. No matter what we choose I think the key points about this API is that it should be stable no matter happens int the compiler while being simple enough to maintain.

I think it's also fine if we don't get it right the first time as long as we get it right by the time we try to get it merged in the compiler.

> I also think we might want to make loc required, and have people manually write Location.none where they want no location, rather than making it easy to forget to pass a location argument.

I totally agree. To be honest I wonder if we shouldn't go even further and force an actual location to be passed. Right now `none` isn't exposed in astlib's `Location`, maybe it should stay that way. There's nothing more frustrating than using a `ppx` that throws `Error at File _ line -1` or whatever `none` points to. That might be a bit extreme though.

NathanReb commented:
> That might be a bit extreme though

And potentially not the best way to achieve that. I think the point is that we should provide an API that encourages ppx authors to properly report errors. I'll make `loc` mandatory!

rgrinberg commented:
In my opinion using formatters should be required and there's no need for a string based version. When reporting errors, users should be equally mindful of formatting as they are of location information. Also, in compiler-libs, the most commonly used function for reporting errors is this one (the `?sub` has been added recently):

```
val raise_errorf: ?loc:t -> ?sub:msg list ->
  ('a, Format.formatter, unit, 'b) format4 -> 'a
```

So ppx authors are already used to using formatters to write error messages. 

Even in ppxlib, the function is:

```
val raise_errorf : ?loc:t -> ('a, Caml.Format.formatter, unit, 'b) format4 -> 'a
```

So it makes most sense to me to just follow this example.

ghost commented:
Just something I thought of since we first discussed this: raising in a ppx rewriter is not great for merlin as it throws away the whole AST even if only a small portion of it is invalid. We started some work at Jane Street to create `[%error ...]` extension points instead of raising in ppx rewriters. These extension points are recognised by the compiler and produce a nice `File ...` error message.

Since `ppx` is new, we might as well do the right thing right from the start. So instead of exposing a function to raise an exception, we should provide a way to produce `[%error ...]` extension points. We should also expose a way to report errors for the case when it is the ppx driver that reports error, as it is the case with Dune. If we want to transport errors through an exception in `ppx`, then we should define this exception in `ppx`.

Regarding string vs formatters, the idea is to expose a generic version in Astlib and have the version for end users live in the `ppx` library. Nathan was mentioning to me yesterday that with recent compilers, the formatting of errors is done late. And this is an important detail as it might be done as part of a larger formatting. As a result, the error creation function in Astlib should indeed take a formatter. It could be either a `(...) format4` or simply a `(Format.formatter -> unit)` callback, to match the type of `~sub`

NathanReb commented:
I updated this PR and now `Location.Error` contains:
- an abstract type and a `make` function to which we can add optional arguments in the future without breaking call sites
- a report function to format error messages given an `Error.t`
- a `to_extension` function that turns an `Error.t` into a `Parsetree.extension`

Once astlib goes into the compiler I have no doubt that `Error.t` will in fact be `Location.report` and that the `report` function will have a proper implementation but given that this implementation has to be compatible with a wide range of OCaml versions so I had to define the type as it is since error representation has changed quite a bit between now and 4.02.

For instance the `sub` errors went from being `error list` to `(Format.formatter -> unit) list` and the `if_highlight` field was removed. The only part of the API that remained stable is the one for generating an error with its loc and main message.

Next step is to had wrappers in `ppx`. Given that I think a good first version would be an `error_extf : loc: Location.t -> ('a, unit, string, extension) format4 -> 'a` that wold return an extension directly it would probably fit into `Ast_builder` so that we can use it in `ppx_view` and `metaquot`.

Let me know what you think!

NathanReb commented:
You'll also note that I removed the `-open Ocaml_shadow` because I needed to use the actual parsetree module. I'm happy to move to a safer alternative but I felt like it was okay since this is ultimately going to end up into the compiler where it will have access to the compiler-libs.

ghost commented:
Well, Astlib will go in the compiler eventually so it can see the compiler modules. It's only what comes after that shouldn't look at the compiler.

diml commented on astlib/location.mli:
```diff
@@ -40,3 +40,19 @@ val update
+  val make :
+    loc: location ->
+    fmt_msg: (Format.formatter -> unit) ->
+    unit ->
```
I suggest this API which is slightly simpler:

```ocaml
val make : loc:location -> (Format.formatter -> unit) -> unit
```

ceastlund replied:
I wouldn't call that `make` if it returns a unit, but perhaps `report`. At which point we don't need a submodule at all, but just `Location.report_error` or some such. I agree that we don't need two steps and an indirection via `Error.t`.

NathanReb replied:
I think the `unit` in @diml suggestion was a typo and that he meant `t`.

Just to clarify you're suggesting we get rid of the abstract type and add arguments to the `report` and `to_extension` functions right?

I agree that right now the indirection isn't necessary but I feel like it's more future proof that way. If we need to add new functions there, we're going to have to duplicate the arguments for each of them and ultimately it's just going to end up like `Location`, a bunch of slightly different functions we try to maintain stable. I feel like an abstract type is a better way to deal with this.
 
Also @diml suggested that at some point we'll need a way to convert from extension points to errors. I'm guessing that's only necessary when we wish to report them so we could add a report function that takes an `extension` as input but at this point it would roughly correspond to using `extension` as an unsafe and exposed type for errors.

Again since it's all going to be wrapped by `ppx` it's not too much of an issue but I'm still not sure that's the best way to go.

diml replied:
That was indeed a typo. The number 1 priority for `Astlib` is that it is easy to preserve the interface when the compiler changes. If the API is a bit clunky to use, that's fine, we'll just add our own nicer API on top.

ghost commented:
Apart from the comment about the arguments, that seems good to me

NathanReb commented:
I fixed the `make` function, added a bit of documentation and fixed the cinaps library which I broke when removing `ocaml-compiler-libs` from `astlib`'s dependencies as it was relying on it as a transitive deps.

We should consider disabling `implicit_transitive_deps`!

ghost commented:
+1 for `implicit_transitive_deps`.

NathanReb commented:
Let's merge!
